### ISSUE TITLE
Refactor `Magento_BundleConfig` module to follow best-practices

### ISSUE THREAD

**Issue by DrewML** [magento/m2-devtools#7]
## This issue is a:

- [ ] Bug
- [ ] Feature suggestion
- [ ] Documentation issue
- [X] Other (Code Quality task)

### Environment

| Question                    |    Answer     |
| --------------------------- | :-----------: |
| Magento version             | N/A |
| Browser + version           | N/A |
| node.js version (`node -v`) | N/A |
| npm version (`npm -v`)      | N/A |

## Description

The [`Magento_BundleConfig`](https://github.com/magento/m2-devtools/tree/268ea4b3f91211b3ce6bd7aac8a74adbcda60b01/Magento_BundleConfig) module was authored by me. I barely know PHP, and I certainly don't know best-practices for m2 modules. The majority of this code is copy/pasted from `Magento_RequireJs` in core.

## Expected result:
`Magento_BundleConfig` is refactored to follow Magento module best-practices.

**Comment by Jakhotiya** [comment]
@DrewML Do we really need this module? I used `bundles` configuration suggested in 
[Advanced Javascript Bundling](https://devdocs.magento.com/guides/v2.3/performance-best-practices/advanced-js-bundling.html) . After a week of using it I find it assuring that even if your bundles aren't covering all the needed javascript files , your site will still work because require has a fallback option. Am I missing a downside of this approach? Is there a upside to installing the module?

**Comment by Vinai** [comment]
@aligent-lturner `setup_version` has been deprecated in Magento 2.3.0.
If the module is intended to work with 2.2 or 2.1 however, it is indeed required, as declarative schema hasn't been backported.

@Jakhotiya I believe the benefit of this plugin is ease and speed of maintenance of the bundle config (`build.js`). Also, this approach will include any custom requirejs modules, that otherwise would have to be included manually in the configuration.

**Comment by Jakhotiya** [comment]
@Vinai You are talking about "Magento_BundleConfig" magento module right? How does that help us in ease of maintenance of the bundle config ? The module just adds appropriate bundled files to head of the page depending on the type of the page. I'm confused, am I misunderstanding?

**Comment by Vinai** [comment]
It is intended to be used with the chrome extension that generates the bundle configuration.
So the chrome extension generates the configuration
Then the requirejs optimizer is used to build the bundles
Then this module adds them to the page.

**Comment by Jakhotiya** [comment]
@Vinai I understand that *Magento_BundleConfig*  adds bundles to the page. I do. I'm saying we don't need to add bundles to page head (which is what this module does.)
Instead we add following config requirejs-config.js

`require.config({ bundles: { "bundles/m2-bootstrap":["a", "b", "c" ] } })`

so when some code requires module "a", requirejs knows that it needs to download  "bundles/m2-bootstrap". Hence only the relevant module is loaded. We don't need to add those bundles to head.

You can configure all of your bundles like that as suggested by the Advanced javascript bundling doc.

I think that reduces overhead of developing and installing an additional magento module. My intention is to ask this question
"Do we need to develop Magento_BundleConfig module?"

**Comment by Vinai** [comment]
I think it's a good idea to start loading the appropriate bundles as `async` assets  before they are required by some other module - the page will be interactive much sooner if the actual loading is not triggered by requirejs.
But maybe I'm missing a point here.

**Comment by Vinai** [comment]
@Jakhotiya If I'm wrong, where do you suggest the bundles configuration should be added to? A theme level `requirejs-config.js` file?

**Comment by Jakhotiya** [comment]
@Vinai Hmmm...You have solid point when use word `async`. I've not thought about adding `async` attribute to those script. I should probably experiment with that. Right now the module doesn't add that attribute to async attribute.  

```

<script  type="text/javascript"  src="https://mage.dev/static/version1544538193/frontend/Magento/luma/en_US/bundles/shared.js"></script>
 <script  type="text/javascript"  src="https://mage.dev/static/version1544538193/frontend/Magento/luma/en_US/bundles/cms-index-index.js"></script>

```
As of now they are render blocking. So that's kinda not good.

*To answer how to add them to requirejs-config.js*
We don't add those config to theme's javascript.  In our build.js file (build configuration) we add this
```javascript
 onModuleBundleComplete: function (data) {
        function onBundleComplete (config, data) {
            const fileName = `${config.dir}/requirejs-config.js`;
            const bundleConfig = {};
            bundleConfig[data.name] = data.included;
            bundleConfig[data.name] = bundleConfig[data.name].map(bundle => bundle.replace(/\.js$/, ''));
            const contents = `require.config({
           bundles: ${JSON.stringify(bundleConfig)},
       });`;
            fs.appendFile(fileName, contents,function(err){
                if(err) console.log(err);
            });
        }
        onBundleComplete(config, data);
    }

```
What this does is, adds those configurations to deployed requirejs-config.js file. All of this is documented in Advanced Javascript bundling doc in Magento devdocs.  @Vinai Can you try this setup out

I should Probably create pull request for that. What do you think @DrewML

**Comment by DrewML** [comment]
@Jakhotiya I would not recommend using the `bundles` configuration option for this. Few reasons:

1. Browsers have a [preload scanner](https://andydavies.me/blog/2013/10/22/how-the-browser-pre-loader-makes-pages-load-faster/) that tokenizes and quickly looks through the HTML document when it's first flushed from the server to begin fetching scripts. If these script tags are loaded by RequireJS configuration, that would mean the browser could not start fetching `shared.js` or `page-handle-here.js` until `require.js` and `requirejs-config.js` have downloaded + executed.

2. There are situations where loading an entire bundle for 1 module would not be ideal. Let's walk through a scenario:

	A simple product page loads `product/foo` and `product/bar` eagerly (in the `define` call), so they are both in the `catalog-product-view.js` bundle. Now imagine that in a JS file for checkout, someone adds a `require` to lazy load `product/foo` module for a certain feature in the UI when clicked. Because of this, checkout will now fetch the _entire_ `catalog-product-view.js` bundle, even though it would have been much faster to just download a copy of `product/foo.js` directly.

3. These scripts are indeed currently blocking. However, putting the `async` attribute on them will likely break functionality in your store, as `async` instructs the browser to _execute_ scripts in any order, as soon as they're downloaded. What you're likely looking for here is `defer`: `defer` will not block render, but will ensure scripts are executed in order. `defer` is currently not added by `Magento_BundleConfig` because there is at least 1 script in a default m2 installation in the body that expects certain modules to be available synchronously. If we got a fix into core we could likely adopt `defer` here.

**Comment by DrewML** [comment]
To keep things organized, I'm going to close this issue out since the work was done by @Vinai in #39. @Jakhotiya if you want to discuss anything more from the thread above re: `bundles` want to throw up a new issue?

**Comment by Jakhotiya** [comment]
Thanks @DrewML Your thoughts on this are food for thought. I admit I'm still very confused about what is the better way. Because in our scenario most of the javascript should execute only after DOM content is loaded. 
I agree async will be trouble. I
 don't see "defer" being a possibility in near future.
 I'm gonna try this module out with http2 server push and see if script really causes problems in first contentful paint on 3G network. Hopefully I find some info to share in next few deployments.

**Comment by DrewML** [comment]
Oh, I forgot to mention there is also a race condition currently with async/defer that can lead to loading duplicate code.

When `requirejs-config.js` executes in the `head` (synchronously), it declares `deps`, which are really just an alias for `require`, and the runtime will schedule the work to start fetching those modules immediately. When that happens, if your bundles haven't executed and registered a module with an identifier that was in `deps`, that module will be fetched as an external dep, and the one in the bundle won't be used.

There is an open question though (haven't run traces) of whether this would be better or worse for perf. Plus we can address the race
